File: Day-10/retry-queue-sample/python/retry_queue.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from typing import Deque, Dict, Generic, TypeVar

T = TypeVar("T")


@dataclass
class Job(Generic[T]):
    job_id: str
    payload: T
    attempts: int = 0


@dataclass
class RetryQueue(Generic[T]):
    max_retries: int
    capacity: int
    _counter: int = field(default=0, init=False)
    _queue: Deque[Job[T]] = field(default_factory=deque, init=False)
    _inflight: Dict[str, Job[T]] = field(default_factory=dict, init=False)
    _dead_letter: list[Job[T]] = field(default_factory=list, init=False)

    def __post_init__(self) -> None:
        if self.max_retries < 0:
            raise ValueError("max_retries must be >= 0")
        if self.capacity <= 0:
            raise ValueError("capacity must be positive")

    def enqueue(self, payload: T) -> str | None:
        if len(self._queue) >= self.capacity:
            return None

        self._counter += 1
        job = Job(job_id=f"job-{self._counter}", payload=payload)
        self._queue.append(job)
        return job.job_id

    def poll(self) -> Job[T] | None:
        if not self._queue:
            return None

        job = self._queue.popleft()
        self._inflight[job.job_id] = job
        return job

    def ack(self, job_id: str) -> bool:
        return self._inflight.pop(job_id, None) is not None

    def fail(self, job_id: str) -> str:
        job = self._inflight.pop(job_id, None)
        if job is None:
            return "missing"

        job.attempts += 1
        if job.attempts <= self.max_retries:
            self._queue.append(job)
            return "requeued"

        self._dead_letter.append(job)
        return "dead-lettered"

    def queue_ids(self) -> list[str]:
        return [job.job_id for job in self._queue]

    def inflight_ids(self) -> list[str]:
        return list(self._inflight.keys())

    def dead_letter_ids(self) -> list[str]:
        return [job.job_id for job in self._dead_letter]

File: Day-10/retry-queue-sample/java/RetryQueue.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class RetryQueue<T> {
    public static final class Job<T> {
        private final String jobId;
        private final T payload;
        private int attempts;

        private Job(String jobId, T payload) {
            this.jobId = jobId;
            this.payload = payload;
            this.attempts = 0;
        }

        public String getJobId() {
            return jobId;
        }

        public T getPayload() {
            return payload;
        }

        private int incrementAttempts() {
            attempts += 1;
            return attempts;
        }
    }

    private final int maxRetries;
    private final int capacity;
    private int counter;
    private final Deque<Job<T>> queue;
    private final Map<String, Job<T>> inflight;
    private final List<Job<T>> deadLetter;

    public RetryQueue(int maxRetries, int capacity) {
        if (maxRetries < 0) {
            throw new IllegalArgumentException("maxRetries must be >= 0");
        }
        if (capacity <= 0) {
            throw new IllegalArgumentException("capacity must be positive");
        }
        this.maxRetries = maxRetries;
        this.capacity = capacity;
        this.counter = 0;
        this.queue = new ArrayDeque<>();
        this.inflight = new HashMap<>();
        this.deadLetter = new ArrayList<>();
    }

    public String enqueue(T payload) {
        if (queue.size() >= capacity) {
            return null;
        }

        counter += 1;
        Job<T> job = new Job<>("job-" + counter, payload);
        queue.addLast(job);
        return job.getJobId();
    }

    public Job<T> poll() {
        Job<T> job = queue.pollFirst();
        if (job == null) {
            return null;
        }
        inflight.put(job.getJobId(), job);
        return job;
    }

    public boolean ack(String jobId) {
        return inflight.remove(jobId) != null;
    }

    public String fail(String jobId) {
        Job<T> job = inflight.remove(jobId);
        if (job == null) {
            return "missing";
        }

        int attempts = job.incrementAttempts();
        if (attempts <= maxRetries) {
            queue.addLast(job);
            return "requeued";
        }

        deadLetter.add(job);
        return "dead-lettered";
    }

    public List<String> queueIds() {
        return queue.stream().map(Job::getJobId).collect(Collectors.toList());
    }

    public List<String> inflightIds() {
        return new ArrayList<>(inflight.keySet());
    }

    public List<String> deadLetterIds() {
        return deadLetter.stream().map(Job::getJobId).collect(Collectors.toList());
    }
}